ycmd/completers/all/identifier_completer.py

```diff
@@ -37,6 +37,10 @@ def __init__( self, user_options ):
     self._completer = ycm_core.IdentifierCompleter()
     self._tags_file_last_mtime = defaultdict( int )
     self._logger = logging.getLogger( __name__ )
+    try:
+      self._max_candidates = user_options[ 'max_num_identifier_candidates' ]
+    except KeyError:
+      self._max_candidates = MAX_IDENTIFIER_COMPLETIONS_RETURNED
```

Comment:
Here is the logic for loading option files: https://github.com/Valloric/ycmd/blob/8e228643ffcab2aa274777ca8b9b0c8183f0ef24/ycmd/__main__.py#L114-L116

Either the option file from commandline argument gets loaded and if not specified, then `default_settings.json`
There is no fallback. You get a `KeyError` if the option is missing.